cdc/cdc.cc

```diff
@@ -0,0 +1,221 @@
+    return table_name + cdc_log_suffix;
+}
+
+sstring desc_name(const sstring& table_name) {
+    static constexpr auto cdc_desc_suffix = "_scylla_cdc_desc";
+    return table_name + cdc_desc_suffix;
+}
+
+static future<>
+remove_log(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    try {
+        return ctx._migration_manager.announce_column_family_drop(
+                ks_name, log_name(table_name), false);
+    } catch (exceptions::configuration_exception& e) {
+        // It's fine if the table does not exist.
+        return make_ready_future<>();
+    }
+}
+
+static future<>
+remove_desc(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    try {
+        return ctx._migration_manager.announce_column_family_drop(
+                ks_name, desc_name(table_name), false);
+    } catch (exceptions::configuration_exception& e) {
+        // It's fine if the table does not exist.
+        return make_ready_future<>();
+    }
+}
+
+future<>
+remove(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    return when_all(remove_log(ctx, ks_name, table_name),
+                    remove_desc(ctx, ks_name, table_name)).discard_result();
+}
+
+static future<> setup_log(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), log_name(s.cf_name()));
+    b.set_default_time_to_live(gc_clock::duration{86400}); // 24h
+    b.set_comment(sprint("CDC log for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("stream_id", uuid_type, column_kind::partition_key);
+    b.with_column("time", timeuuid_type, column_kind::clustering_key);
+    b.with_column("batch_seq_no", int32_type, column_kind::clustering_key);
+    b.with_column("operation", int32_type);
+    b.with_column("ttl", long_type);
+    auto add_columns = [&] (const auto& columns) {
+        for (const auto& column : columns) {
+            b.with_column("_" + column.name(), column.type);
+        }
+    };
+    add_columns(s.partition_key_columns());
+    add_columns(s.clustering_key_columns());
+    add_columns(s.static_columns());
+    add_columns(s.regular_columns());
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+static future<> setup_desc(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), desc_name(s.cf_name()));
+    b.set_comment(sprint("CDC description for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("node_ip", inet_addr_type, column_kind::partition_key);
+    b.with_column("shard_id", int32_type, column_kind::partition_key);
+    b.with_column("created_at", timestamp_type, column_kind::clustering_key);
+    b.with_column("stream_id", uuid_type);
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+template <typename Func>
+static void for_each_endpoint(db_context ctx, Func&& f) {
+    for (const auto& dc : ctx._token_metadata.get_topology().get_datacenter_endpoints()) {
```

Comment:
Please use std::function instead of a template. This both helps getting better error messages, and reduces code bloat.

Reply:
I removed those helper at all.